src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -115,105 +113,147 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
 
-                // remove projects that exist in both side
-                olds.Object.ExceptWith(newSolutionChecksums.Projects.Checksums);
-                news.Object.ExceptWith(oldSolutionChecksums.Projects.Checksums);
+                foreach (var (oldChecksum, projectId) in oldSolutionChecksums.Projects)
+                    oldProjectIdToChecksum.Add(projectId, oldChecksum);
 
-                using var _1 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var oldMap);
-                using var _2 = PooledDictionary<ProjectId, ProjectStateChecksums>.GetInstance(out var newMap);
+                foreach (var (newChecksum, projectId) in newSolutionChecksums.Projects)
+                    newProjectIdToChecksum.Add(projectId, newChecksum);
 
-                await PopulateOldProjectMapAsync().ConfigureAwait(false);
-                await PopulateNewProjectMapAsync(this).ConfigureAwait(false);
+                allProjectIds.AddRange(oldSolutionChecksums.Projects.Ids);
+                allProjectIds.AddRange(newSolutionChecksums.Projects.Ids);
 
-                // bulk sync assets
-                await SynchronizeAssetsAsync(oldMap, newMap, cancellationToken).ConfigureAwait(false);
-
-                // added project
-                foreach (var (projectId, newProjectChecksums) in newMap)
+                // If there are old projects that are now missing on the new side, and this is a projectConeSync, then
+                // exclude them from the old side as well.  This way we only consider projects actually added or
+                // changed.
+                //
+                // Importantly, this means in the event of a cone-sync, we never drop projects locally.  That's very
+                // desirable as it will likely be useful in future calls to still know about that project info without
+                // it being dropped and having to be resynced.
+                var isConeSync = newSolutionChecksums.ProjectConeId != null;
+                if (isConeSync)
                 {
-                    if (!oldMap.ContainsKey(projectId))
+                    foreach (var (oldChecksum, oldProjectId) in oldSolutionChecksums.Projects)
                     {
-                        var projectInfo = await _assetProvider.CreateProjectInfoAsync(projectId, newProjectChecksums.Checksum, cancellationToken).ConfigureAwait(false);
-                        solution = solution.AddProject(projectInfo);
+                        if (!newProjectIdToChecksum.ContainsKey(oldProjectId))
+                            oldProjectIdToChecksum.Remove(oldProjectId);
                     }
+
+                    // All the old projects must be in the new project set.  Though the reverse doesn't have to hold.
+                    // The new project set may contain additional projects to add.
+                    Debug.Assert(oldProjectIdToChecksum.Keys.All(newProjectIdToChecksum.Keys.Contains));
                 }
```

Comment:
the above is the meat of the change.  effectively, with a cone sync, on the remote-side, we say "we'll assume everything outside of the cone being synced is not present on the remote side, so we won't touch that at all.